views/default/page/components/table.php

```diff
@@ -0,0 +1,130 @@
+ * @uses $vars['pagination']   Show pagination? (default: true)
+ * @uses $vars['position']     Position of the pagination: before, after, or both
+ * @uses $vars['full_view']    Show the full view of the items (default: false)
+ * @uses $vars['list_class']   Additional CSS class for the <table> element
+ * @uses $vars['item_class']   Additional CSS class for the <td> elements
+ * @uses $vars['item_view']    Alternative view to render list items
+ * @uses $vars['no_results']   Message to display if no results (string|Closure)
+ */
+$items = $vars['items'];
+$count = elgg_extract('count', $vars);
+$pagination = elgg_extract('pagination', $vars, true);
+$position = elgg_extract('position', $vars, 'after');
+$no_results = elgg_extract('no_results', $vars, '');
+$cell_views = elgg_extract('cell_views', $vars, ['page/components/table/cell/default']);
+$headings = elgg_extract('headings', $vars, '');
+
+if (empty($cell_views) || !is_array($cell_views)) {
+	elgg_log('$vars["cell_views"] must be an array of view names', 'ERROR');
+	return;
+}
+
+if (is_array($headings)) {
+	if (count($headings) !== count($cell_views)) {
+		// complain but carry on rendering
+		elgg_log('$vars["headings"] size must match $vars["cell_views"]', 'ERROR');
+		$headings = '';
+	} else {
+		$heading_renderer = function ($heading) {
+			return "<th>$heading</th>";
+		};
+		$headings = array_map($heading_renderer, $headings);
+		$headings = implode('', $headings);
+		$headings = "<thead><tr>$headings</tr></thead>";
+	}
+}
+
+if (!$items && $no_results) {
+	if ($no_results instanceof Closure) {
+		echo $no_results();
+		return;
+	}
+	echo "<p class='elgg-no-results'>$no_results</p>";
+	return;
+}
+
+if (!is_array($items) || count($items) == 0) {
+	return;
+}
+
+$table_classes = ['elgg-list', 'elgg-table'];
+if (isset($vars['list_class'])) {
+	$table_classes[] = $vars['list_class'];
+}
+
+$cell_classes = ['elgg-item'];
+if (isset($vars['item_class'])) {
+	$cell_classes[] = $vars['item_class'];
+}
+
+$nav = ($pagination) ? elgg_view('navigation/pagination', $vars) : '';
+
+$rows = '';
+foreach ($items as $item) {
+	$cell_attrs = ['class' => $cell_classes];
+	$row_attrs = [];
+
+	if ($item instanceof \ElggEntity) {
+		$guid = $item->getGUID();
+		$type = $item->getType();
+		$subtype = $item->getSubtype();
+
+		$cell_attrs['id'] = "elgg-$type-$guid";
+		$row_attrs['id'] = "elgg-row-$type-$guid";
+
+		$cell_attrs['class'][] = "elgg-item-$type";
+		if ($subtype) {
+			$cell_attrs['class'][] = "elgg-item-$type-$subtype";
+		}
+	} else if (is_callable(array($item, 'getType'))) {
+		$cell_attrs['id'] = "item-{$item->getType()}-{$item->id}";
+		$row_attrs['id'] = "item-row-{$item->getType()}-{$item->id}";
+	}
+
+	$row = '';
+
+	foreach ($cell_views as $cell_view) {
+		$cell = elgg_view($cell_view, [
+			'item' => $item,
+			'item_vars' => $vars,
+			'attrs' => $cell_attrs,
+		]);
+		if (!$cell) {
+			// abandon the row if can't draw one cell
```

Comment:
Agreed. In the list view we always render the item and if its falsy we cancel the row. I guess we'll just be allowing empty cells in that case now.